lib/streamlit/runtime/caching/cache_utils.py

```diff
@@ -52,22 +57,33 @@
 )
 from .hashing import update_hash
 
-if TYPE_CHECKING:
-    from streamlit.delta_generator import DeltaGenerator
-
 _LOGGER = get_logger(__name__)
 
 
+@dataclass
+class WidgetMsgMetadata:
+    """Everything needed for replaying a widget and treating it as an implicit
+    argument to a cached function, beyond what is stored for all elements.
+    """
+
+    widget_id: str
+    widget_value: Any
+    metadata: WidgetMetadata[Any]
```

Comment:
Made some of them frozen, which I think is strictly better than converting to NamedTuple. Left `CachedResult` and `InitialCachedResult` as is, to indicate mutability, even though they have internal mutability (mutation of their fields) rather than external (mutating what their fields point to)